### PR TITLE
Fix and enable dialyzer

### PR DESCRIPTION
**Pull request by ukutaht**
### Changes

Fixed dialyzer errors and enabled it in CI.

I had to mark one function as `:nowarn` because we use an ecto feature that does not have a proper typespec.

### CODE REVIEW

File: lib/plausible_web/controllers/stats_controller.ex [hunk 1]
```diff
@@ -217,12 +218,13 @@ defmodule PlausibleWeb.StatsController do
 
   defp find_shared_link(domain, auth) do
     link_query =
-      from link in Plausible.Site.SharedLink,
```
Comment (ukutaht) [inline lib/plausible_web/controllers/stats_controller.ex:220]: Sorry I think my editor added some formatting changes as well

**Comment by ukutaht** [comment]
What do we think about Credo? 

> 1396 mods/funs, found 2 consistency issues, 26 refactoring opportunities, 114 code readability issues, 2 software design suggestions.

**Comment by ukutaht** [comment]
> What do we think about Credo?
> 
> > 1396 mods/funs, found 2 consistency issues, 26 refactoring opportunities, 114 code readability issues, 2 software design suggestions.

The issues are quite simple, we could go down the list, fix them and then enable Credo in CI as well

**Comment by vinibrsl** [comment]
> What do we think about Credo?
> 
> > 1396 mods/funs, found 2 consistency issues, 26 refactoring opportunities, 114 code readability issues, 2 software design suggestions.

Credo would be a nice addition to our CI pipeline. It would require fixing current issues I think, but helps us maintaining good Elixir conventions and style consistency.

**Comment by ukutaht** [comment]
> helps us maintaining good Elixir conventions and style consistency.

For me the biggest value is taking some discussions / arguments / disagreements off the table. I find dev teams spend a lot of time bikeshedding over code style which is down to personal preference. Having an automated rule system means we don't need to spend time arguing about style and can spend more energy on the substance.

**Comment by ukutaht** [comment]
@ruslandoga has opened a PR on Ecto that will allow us to remove the `:nowarn` hint https://github.com/elixir-ecto/ecto/pull/4026 ❤️

File: lib/plausible_web/controllers/stats_controller.ex (hunk 1 above)
Comment (aerosol) [reply lib/plausible_web/controllers/stats_controller.ex:220]: `mix format` should use our local `.formatter.exs` where we import `:ecto` and where all the macros as listed (https://github.com/elixir-ecto/ecto/blob/master/.formatter.exs)

**Comment by ukutaht** [comment]
Credo has a diff mode so we can at least fail the build when new issues are introduced: https://github.com/rrrene/credo/blob/master/CHANGELOG.md#new-diff-options

This just ensures no new issues are introduced in a PR but it won't force us to fix any of the existing ones. Better than nothing but it won't force us to fix old problems along the way.

I'm not sure how to get it to do a full run on only the changed files. I tried the following without success:
```bash
$ mix credo suggest --files-included $(git diff --name-only master HEAD)
```

**Comment by ukutaht** [comment]
```
➜  analytics git:(fix-dialyzer) mix credo suggest --files-included $(git diff --name-only master HEAD)
dependency :hut is using Rebar 2, which is no longer maintained and no longer works in recent Erlang/OTP versions. Remove the :manager option or set it to :rebar3 instead
info: Some source files could not be parsed correctly and are excluded:

   1) .github/workflows/elixir.yml

No files found!

Please report incorrect results: https://github.com/rrrene/credo/issues

Analysis took 0.05 seconds (0.04s to load, 0.01s running 52 checks on 0 files)
0 mods/funs, found no issues.

Showing priority issues: ↑ ↗ →  (use `mix credo explain` to explain issues, `mix credo --help` for options).
```

**Review by vinibrsl — APPROVED** [review]

**Comment by vinibrsl** [comment]
I added two commits that runs Credo against Git changes only. CI is green and happy :)

File: .github/workflows/elixir.yml [hunk 2]
```diff
@@ -66,11 +68,11 @@ jobs:
         run: mix compile --warnings-as-errors --all-warnings
       - name: Generate tracker files
         run: npm run deploy --prefix ./tracker
+      - name: Check Credo Warnings
+        run: mix credo diff --from-git-merge-base origin/master
```
Comment (aerosol) [inline .github/workflows/elixir.yml:72]: is this expected to halt the CI with error?

File: .github/workflows/elixir.yml [hunk 3]
```diff
@@ -40,7 +40,9 @@ jobs:
             --health-retries 5
 
     steps:
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@v1
```
Comment (aerosol) [inline .github/workflows/elixir.yml:43]: Why has this version had to change?

File: .github/workflows/elixir.yml (hunk 2 above)
Comment (vinibrsl) [reply .github/workflows/elixir.yml:72]: It halts CI with error with the exit code.

File: .github/workflows/elixir.yml (hunk 3 above)
Comment (vinibrsl) [reply .github/workflows/elixir.yml]: Previous commits were using Credo piping through some Git magic to list changed files. I changed the Credo command to `mix credo diff --from-git-merge-base origin/master` which does a different Git magic that does not need `checkout@v1`. I also updated actions/checkout to the latest version.

**Pull request merged**